cluster/kubernetes/namespacer.go

```diff
@@ -0,0 +1,166 @@
+package kubernetes
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	crdv1beta1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+	crd "k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/discovery"
+	discocache "k8s.io/client-go/discovery/cached"
+	toolscache "k8s.io/client-go/tools/cache"
+
+	kresource "github.com/weaveworks/flux/cluster/kubernetes/resource"
+)
+
+type namespaceViaDiscovery struct {
+	fallbackNamespace string
+	disco             discovery.DiscoveryInterface
+}
+
+type namespaceDefaulter interface {
+	GetDefaultNamespace() (string, error)
+}
+
+// NewNamespacer creates an implementation of Namespacer
+func NewNamespacer(ns namespaceDefaulter, d discovery.DiscoveryInterface) (*namespaceViaDiscovery, error) {
+	fallback, err := ns.GetDefaultNamespace()
+	if err != nil {
+		return nil, err
+	}
+	return &namespaceViaDiscovery{fallbackNamespace: fallback, disco: d}, nil
+}
+
+// effectiveNamespace yields the namespace that would be used for this
+// resource were it applied, taking into account the kind of the
+// resource, and local configuration.
+func (n *namespaceViaDiscovery) EffectiveNamespace(m kresource.KubeManifest) (string, error) {
+	namespaced, err := n.lookupNamespaced(m.GroupVersion(), m.GetKind())
+	switch {
+	case err != nil:
+		return "", err
+	case namespaced && m.GetNamespace() == "":
+		return n.fallbackNamespace, nil
+	case !namespaced:
+		return "", nil
+	}
+	return m.GetNamespace(), nil
+}
+
+func (n *namespaceViaDiscovery) lookupNamespaced(groupVersion, kind string) (bool, error) {
+	resourceList, err := n.disco.ServerResourcesForGroupVersion(groupVersion)
+	if err != nil {
+		return false, fmt.Errorf("error looking up API resources for %s.%s: %s", kind, groupVersion, err.Error())
+	}
+	for _, resource := range resourceList.APIResources {
+		if resource.Kind == kind {
+			return resource.Namespaced, nil
+		}
+	}
+	return false, fmt.Errorf("resource not found for API %s, kind %s", groupVersion, kind)
+}
+
+// This mainly exists so I can put the controller and store in here,
+// so as to have access to them for testing; and, so that we can do
```

Comment:
Is there a reason not to move `store` and `controller` to the tests (as local variables initialized after calling `MakeCachedDiscovery`? I think that placing them here makes it more difficult to read the code.

Reply:
The fields are assigned during construction, and wouldn't be kept except that it's useful during testing to be able to reach in and check what's been recorded. Could you elaborate on what you're suggesting?

Reply:
Uhm, I am also seeing the `chainHandler` thing, which is currently only used in testing. If you are not planning to use it with normal code, I would also move it to testing.

For instance we could create a private `makeCacheDiscovery` which the tests could use:

```golang
func MakeCachedDiscovery(d discovery.DiscoveryInterface, c crd.Interface, shutdown <-chan struct{}) *cachedDiscovery {
  lw, handler, result := makeCachedDiscovery(d, c, shutdown)
  toolscache.NewInformer(lw, &crdv1beta1.CustomResourceDefinition{}, 5*time.Minute, handler)
  return result
}
```

The tests could invoke `makeCachedDiscovery` and then call `toolscache.NewInformer` after chaining the handles and keep the results locally.

It's not as pretty as I would like to, so I will lave it up to you.

Reply:
OK, I get the suggestion now -- thanks Fons